### PR TITLE
[Arrow] HugeInt, UHugeInt, Bit, Time_TZ extensions

### PR DESCRIPTION
**Pull request by pdet**
Hugeint is partially supported in the Arrow integration by defaulting to a `decimal(38,0)`. However, this can cause conversion errors and does not allow for proper round-tripping of this type.

To support both hugeint and uhugeint in Arrow, we export them using the `duckdb.hugeint` and `duckdb.uhugeint` extension names and use a fixed-size blob of 16 bytes as their internal format.

The disadvantage of this approach is that producers/consumers will have to implement our specific types.

In the future, we may want to revert to using decimals when we adopt the `VARINT` type as a decimal physical type.

### CODE REVIEW

**Comment by carlopi** [comment]
Great! I think this is very useful would solve some outstanding issues in duckdb-wasm side where those types can now then be properly handled.

**Comment by jraymakers** [comment]
I'm excited that this should address the long-standing duckdb-wasm issue @carlopi mentioned (with UHUGEINT), and also make it possible to distinguish HUGEINT from DECIMAL.

I'm also happy to see that UUID and JSON now have metadata, which will make it easier to distinguish those types (from HUGEINT/UHUGEINT or VARCHAR).

The one remaining ambiguity I'm aware of is between BLOB and BIT. Would it be possible to add metadata to distinguish those types from each other?

**Comment by jraymakers** [comment]
I think there's another ambiguity between TIME and TIME_TZ. Metadata to distinguish those two would be helpful as well.

**Comment by pdet** [comment]
> I think there's another ambiguity between TIME and TIME_TZ. Metadata to distinguish those two would be helpful as well.

I've sneaked `duckdb.bit` and `duckdb.time_tz` in :-).

**Comment by jraymakers** [comment]
> I've sneaked `duckdb.bit` and `duckdb.time_tz` in :-).

Awesome, thanks!

File: test/arrow/arrow_roundtrip.cpp [hunk 1]
```diff
@@ -90,8 +90,21 @@ TEST_CASE("Test Arrow Extension Types", "[arrow][.]") {
 	// UUID
 	TestArrowRoundtrip("SELECT '2d89ebe6-1e13-47e5-803a-b81c87660b66'::UUID str FROM range(5) tbl(i)");
 
-	// JSON
-	TestArrowRoundtrip("SELECT '{\"name\":\"Pedro\", \"age\":28, \"car\":\"VW Fox\"}'::JSON str FROM range(5) tbl(i)");
+	// // JSON
+	// TestArrowRoundtrip("SELECT '{\"name\":\"Pedro\", \"age\":28, \"car\":\"VW Fox\"}'::JSON str FROM range(5)
```
Comment (Mytherin) [inline test/arrow/arrow_roundtrip.cpp:94]: Is this commented out on purpose?

File: src/optimizer/pushdown/pushdown_get.cpp [hunk 2]
```diff
@@ -6,12 +6,71 @@
...
+	return true;
+}
 
 unique_ptr<LogicalOperator> FilterPushdown::PushdownGet(unique_ptr<LogicalOperator> op) {
 	D_ASSERT(op->type == LogicalOperatorType::LOGICAL_GET);
 	auto &get = op->Cast<LogicalGet>();
 
 	if (get.function.pushdown_complex_filter || get.function.filter_pushdown) {
+		if (get.function.name == "arrow_scan") {
+			if (!CanArrowPushdown(get.returned_types)) {
```
Comment (Mytherin) [inline src/optimizer/pushdown/pushdown_get.cpp:70]: Hard-coding this for a specific table function is not very pretty - should we perhaps expand the `TableFunction` with a callback for this instead? Perhaps this can be moved to a separate PR entirely since it seems unrelated?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR!

I'm not opposed to sneaking this in - but this seems to break existing functionality in its current form, i.e. you are currently able to read `time_tz`, `hugeint` and `bit` columns as regular Arrow types, and after this PR they are converted to extension types instead.

I can see how this is useful for clients like WASM, that use Arrow as an intermediate format internally, but I think this limits usability of these types unnecessarily in other scenarios (e.g. in Python). 

Perhaps we can instead move this behind a flag similar to the arrow offset size/string view/etc?

**Comment by pdet** [comment]
> Thanks for the PR!
> 
> I'm not opposed to sneaking this in - but this seems to break existing functionality in its current form, i.e. you are currently able to read `time_tz`, `hugeint` and `bit` columns as regular Arrow types, and after this PR they are converted to extension types instead.
> 
> I can see how this is useful for clients like WASM, that use Arrow as an intermediate format internally, but I think this limits usability of these types unnecessarily in other scenarios (e.g. in Python).
> 
> Perhaps we can instead move this behind a flag similar to the arrow offset size/string view/etc?

I think that is fine for `bit` and `time_tz`, but hugeint`s previous conversion was a broken hack since it couldn't convert all of its range.

**Comment by Mytherin** [comment]
I would leave the behavior as-is for hugeint as well - given that we want to turn it into a `decimal(39,0)` in the future, right?

I think it's rather confusing from a user perspective to go from `decimal(38,0)` in this release, to a custom extension type `duckdb.hugeint` in v1.1, to `decimal(39,0)` in v1.2.

**Comment by pdet** [comment]
> I would leave the behavior as-is for hugeint as well - given that we want to turn it into a `decimal(39,0)` in the future, right?
> 
> I think it's rather confusing from a user perspective to go from `decimal(38,0)` in this release, to a custom extension type `duckdb.hugeint` in v1.1, to `decimal(39,0)` in v1.2.

Fair enough, I'll minimize the switch-aroo. 

*Bikeshedding:* Any suggestions for what the flag name? `use_duck_extensions`, `total_round_trip`?

**Comment by Mytherin** [comment]
`lossless_conversion` maybe?

**Comment by carlopi** [comment]
Just to add some context: I had tried to come up with an extensibility mechanism for extensions to register custom handling for logical types.

Idea would be that once that is in place, all this logic can be moved to an extension (and iterated independently from duckdb / have less clear stability needs), that then can be loaded by whoever would need this for some reason.

I was not able to come up with something I was confident with, expecially given time constraints, but unsure if that could be considered at some other point.

Idea would be that certain API / situation might need to have do with specific arrow versions / arrow clients, and allowing customizing the transformation would allow to work around potential problems (obviously at the cost of losing wider interoperability).

**Comment by pdet** [comment]
> Just to add some context: I had tried to come up with an extensibility mechanism for extensions to register custom handling for logical types.
> 
> Idea would be that once that is in place, all this logic can be moved to an extension (and iterated independently from duckdb / have less clear stability needs), that then can be loaded by whoever would need this for some reason.
> 
> I was not able to come up with something I was confident with, expecially given time constraints, but unsure if that could be considered at some other point.
> 
> Idea would be that certain API / situation might need to have do with specific arrow versions / arrow clients, and allowing customizing the transformation would allow to work around potential problems (obviously at the cost of losing wider interoperability).
> 
> Thanks @pdet for taking this up.

Yeah, I think we will eventually add some kind of call-back to the Arrow scanner to deal with types that are relevant only to extensions but that are passed through Arrow (e.g., geo).

**Comment by Mytherin** [comment]
> Just to add some context: I had tried to come up with an extensibility mechanism for extensions to register custom handling for logical types.
> 
> Idea would be that once that is in place, all this logic can be moved to an extension (and iterated independently from duckdb / have less clear stability needs), that then can be loaded by whoever would need this for some reason.
> 
> I was not able to come up with something I was confident with, expecially given time constraints, but unsure if that could be considered at some other point.
> 
> Idea would be that certain API / situation might need to have do with specific arrow versions / arrow clients, and allowing customizing the transformation would allow to work around potential problems (obviously at the cost of losing wider interoperability).
> 
> Thanks @pdet for taking this up.

For WASM at least, I think a more logical step forward is to replace Arrow in its internals, and directly use our own data chunks/vectors similar to how all of our other APIs work. We are now doing a lot of extra work to work around the fact that there are subtle differences between Arrow and our own types/internal representation. This is not useful for the Arrow integration because Arrow is intended as a *general purpose* interchange format. The work-arounds we are adding here are only useful because we are **both** the producer and the consumer.

We can still (optionally) offer Arrow as a user-exposed interface at the API boundary in WASM for clients that want to use it - but I don't see the value-add of using Arrow if we are going to produce a "heavily customized Arrow" that is filled with extension types to allow for lossless conversion. At that point why not just use our own chunks/types directly?

**Comment by pdet** [comment]
> > Just to add some context: I had tried to come up with an extensibility mechanism for extensions to register custom handling for logical types.
> > Idea would be that once that is in place, all this logic can be moved to an extension (and iterated independently from duckdb / have less clear stability needs), that then can be loaded by whoever would need this for some reason.
> > I was not able to come up with something I was confident with, expecially given time constraints, but unsure if that could be considered at some other point.
> > Idea would be that certain API / situation might need to have do with specific arrow versions / arrow clients, and allowing customizing the transformation would allow to work around potential problems (obviously at the cost of losing wider interoperability).
> > Thanks @pdet for taking this up.
> 
> For WASM at least, I think a more logical step forward is to replace Arrow in its internals, and directly use our own data chunks/vectors similar to how all of our other APIs work. We are now doing a lot of extra work to work around the fact that there are subtle differences between Arrow and our own types/internal representation. This is not useful for the Arrow integration because Arrow is intended as a _general purpose_ interchange format. The work-arounds we are adding here are only useful because we are **both** the producer and the consumer.
> 
> We can still (optionally) offer Arrow as a user-exposed interface at the API boundary in WASM for clients that want to use it - but I don't see the value-add of using Arrow if we are going to produce a "heavily customized Arrow" that is filled with extension types to allow for lossless conversion. At that point why not just use our own chunks/types directly?

Indeed, but on the other hand, I think our native types are pretty common for database interchanges. I could envision our arrow type extensions becoming a sufficiently generic typing system for data interchange between DBMSs

File: src/common/arrow/arrow_appender.cpp [hunk 3]
```diff
@@ -142,9 +142,14 @@ static void InitializeFunctionPointers(ArrowAppendData &append_data, const Logic
 	case LogicalTypeId::INTEGER:
 		InitializeAppenderForType<ArrowScalarData<int32_t>>(append_data);
 		break;
-	case LogicalTypeId::TIME_TZ:
-		InitializeAppenderForType<ArrowScalarData<int64_t, dtime_tz_t, ArrowTimeTzConverter>>(append_data);
+	case LogicalTypeId::TIME_TZ: {
+		if (append_data.options.lossless_conversion) {
+			InitializeAppenderForType<ArrowScalarData<int64_t>>(append_data);
```
[the remark this answers is not in the document]
Comment (jraymakers) [reply src/common/arrow/arrow_appender.cpp:147]: Interesting that the physical type of `TIME_TZ` is `INT64`, but the `bits` field of `dtime_tz_t` is a [uint64_t](https://github.com/duckdb/duckdb/blob/5bdb091a5d67460da3ca3a89f21b7cdc588d4544/src/include/duckdb/common/types/datetime.hpp#L98). Maybe @hawkfish can shed some light on why?

**Comment by Mytherin** [comment]
Thanks!

File: src/common/arrow/arrow_appender.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (hawkfish) [reply src/common/arrow/arrow_appender.cpp:147]: Not sure. `TIME` predates me, so you would have to ask @Mytherin or @hannes . I created the `dtime_tz_t` type when it became clear we needed a new type, and I made it unsigned because it only has positive values. Plus it had internal bit structure, which is always easier with unsigned types. 

Postgres uses a struct with the time and offset values separated, but that seemed like overkill and would require a new physical type. Unfortunately, I got the ordering semantics wrong and you can't sort them as integers but have to use the struct's comparison methods.
